[FIX] odoo-edi: Move cancellation email now displays correct values

caleb-shelton opened the pull request:
The quantity has been changed to quantity cancelled and the
"Cancelled By" column now has the correct user who cancelled the move.

User-story: 2168

Signed-off-by: Caleb Shelton <caleb.shelton@unipart.io>

jrbirch99 commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,13 @@ class StockMove(models.Model):
+
+    def _action_cancel(self):
+        User = self.env["res.users"]
+        self.cancelled_by = User.browse(self.env.real_uid)
```
nit: I think there should be a space between these two lines so like: 
```suggestion
        User = self.env["res.users"]

        self.cancelled_by = User.browse(self.env.real_uid)
```

PeterAlabaster replied:
second nit: might be able to change to just
```python
self.cancelled_by = self.env.real_uid
```

caleb-shelton replied:
I believe this returned an int. I could be mistaken though

PeterC10 replied:
Yeah it is an int.

So in env you have `user` and `uid`, the UDES 11 version of Odoo only added in the ID of the real user.

On that note though... `real_uid` isn't a thing in vanilla Odoo... might need some thinking about.

PeterC10 replied:
This will need reviewing.

PeterC10 replied:
Thanks for sorting @jrbirch99 - just took a while to think of the nicest/least worse solution! Looks good just one nit.

jrbirch99 commented on addons/edi_stock/data/edi_move_cancellation_data.xml:
```diff
@@ -48,7 +48,7 @@
                 <div style="display:table-row;">
                     <div style="display:table-cell; padding: 0px 5px;">Part Number</div>
-                    <div style="display:table-cell; padding: 0px 5px;">Quantity Ordered</div>
+                    <div style="display:table-cell; padding: 0px 5px;">Quantity</div>
```
In the story it mentions " Change ‘Quantity Ordered’ to ‘Quantity Cancelled’ ", should this be "Quantity Cancelled" instead of "Quantity"?

caleb-shelton replied:
I originally wrote this but thought it was better this way as it was implied with the context.

PeterC10 commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,13 @@ class StockMove(models.Model):
         index=True,
     )
 
+    cancelled_by = fields.Many2one("res.users", string="Cancelled By", readonly=True)
```
Can we make this `cancelled_by_id`? Convention (although not always followed) is to end many2one fields with `_id`

jrbirch99 replied:
Changed

jrbirch99 commented on addons/edi_stock/data/edi_move_cancellation_data.xml:
```diff
@@ -57,13 +57,9 @@
                         <div style="display:table-cell; padding: 0px 5px;">${move.product_id.name_get()[0][1]}</div>
                         <div style="display:table-cell; padding: 0px 5px;">
-                            % if move.sale_line_id:
-                                ${move.sale_line_id.product_uom_qty}
```
Q: Why have we removed this? is it just to simplify?

caleb-shelton replied:
This was removed as this assigns the quantity ordered rather than cancelled.

PeterAlabaster replied:
We might want to check the original story / sign off, as i'm pretty sure product requested this functionality originally?

jrbirch99 replied:
I think these should be the same thing for most cases. Not sure how the old code would work if you had back orders and you cancelled a back order, but I think your change would only show the qty cancelled in the backorder, which is correct.

PeterC10 commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,13 @@ class StockMove(models.Model):
 
+    cancelled_by = fields.Many2one("res.users", string="Cancelled By", readonly=True)
+
+    def _action_cancel(self):
```
Can we add a doc-string?

jrbirch99 replied:
changed

PeterAlabaster commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,13 @@ class StockMove(models.Model):
         index=True,
     )
 
+    cancelled_by = fields.Many2one("res.users", string="Cancelled By", readonly=True)
```
would rather this be named `cancelled_by_id` to conform to m2o naming, having `_id` suffix helps identify it is a relational field

PeterAlabaster replied:
Also, unsure if this along with the `_action_cancel()` hook should / could live elsewhere, it seems like it could be a generic feature, but difficult as this repo likes to be standalone.

jrbirch99 replied:
I am finishing off this story. I think we should keep the `_action_cancel()` hook here as it is only used for EDI at the moment. If in the future if hss want to show the person who cancelled the picking on the pick then maybe we could refactor this and move it elsewhere?

PeterC10 commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,13 @@ class StockMove(models.Model):
+
+    def _action_cancel(self):
+        User = self.env["res.users"]
+        self.cancelled_by = User.browse(self.env.real_uid)
```
Need to make sure this works when self contains multiple moves. Probably safer to use `write`.

caleb-shelton replied:
Agreed. Aight this is my last comment before an 11 hour flight. Gl hf.

jrbirch99 replied:
changed

jrbirch99 committed:
[FIX] odoo-edi: Move cancellation email now displays correct values

- The quantity has been changed to quantity cancelled and the "Cancelled By" column now has the correct user who cancelled the move.
- A unit test was added to ensure partial quantities are reported correctly

User-story: 2168

Signed-off-by: Caleb Shelton <caleb.shelton@unipart.io>

PeterAlabaster commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,19 @@ class StockMove(models.Model):
+        """
+        User = self.env["res.users"]
+
+        self.write({"cancelled_by_id": User.browse(self.env.real_uid).id})
```
when using `.write()` it's fine to use int id, the `browse().id` counteracts itself :)
```suggestion
        self.write({"cancelled_by_id": self.env.real_uid})
```

jrbirch99 replied:
yep :sweat_smile: missed that, should be changed now :smile:

PeterC10 commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,19 @@ class StockMove(models.Model):
+        User = self.env["res.users"]
+
+        self.write({"cancelled_by_id": User.browse(self.env.real_uid).id})
+        super(StockMove, self)._action_cancel()
```
Nit: Can just do `super()._action_cancel()`.

jrbirch99 replied:
changed

PeterC10 commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,19 @@ class StockMove(models.Model):
+
+    def _action_cancel(self):
+        """
+        Set the cancelled_by_id to be the id of the user who cancels the move. This is 
```
Adding a comment here as don't want to hijack the other thread.

But has this been tested when `self` contains multiple moves?

PeterC10 replied:
Might need to do `for move in self:` etc

jrbirch99 replied:
So I had multiple moves and the cancelled_by_id field on all moves got updated with the correct user. 

https://github.com/unipartdigital/odoo/blob/udes-11.0/odoo/models.py#:~:text=Updates%20all%20records,the%20provided%20values.

(hope the link works) From the write method docstring I think it is designed to update all records in the set, so I think it is okay.

PeterC10 replied:
Yes sorry, I missed that you changed this to write so got mixed up with the previous version that was updating the records via dot notation. All good.

ws-mps commented on addons/edi_stock/models/edi_move_cancellation_report_document.py:
```diff
@@ -20,6 +20,19 @@ class StockMove(models.Model):
         index=True,
     )
 
+    cancelled_by_id = fields.Many2one("res.users", string="Cancelled By", readonly=True)
```
Can we add add index=False please?

PeterC10 commented on addons/edi_stock/models/res_users.py:
```diff
@@ -0,0 +1,11 @@
+        """
+        Returns the current user id.  
+        """
+        return self.env.user.id
```
Super nit: Can just do:

`return self.env.uid`
